docs/internal/rfc/009rfc-transition-explorer.md

```diff
@@ -0,0 +1,459 @@
+in [ADR 7][].
+
+[ADR 7]: https://github.com/informalsystems/apalache/tree/unstable/docs/src/adr/007adr-restructuring.md
+
+#### API
+
+The following is a rough sketch of the proposed API for the transition explorer.
+It aims to present a highly abstracted interface, but in terms of existing data
+structures. Naturally, refinements and alterations are to be expected during
+implementation.
+
+We refer to symbolic states as `Frames`, which are understood as a set of
+constraints, and we put this terminology in the API in order to help users
+understand when they should be thinking in terms of constraint sets as opposed
+to concrete states. Concrete states can be obtained by the functions suffixed
+with `Example`.
+
+In essence, this proposed API is only a thin wrapper around the
+[TransitionExecutor
+class](https://github.com/informalsystems/apalache/tree/master/tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/trex/TransitionExecutor.scala).
+During previous iterations of the proposed API we discussed exposing a
+higher-level API, targeted at meeting the requirements more directly.  However,
+discussion revealed that the expensive computational costs of SAT solving in
+most cases made it infeasible to meet the requirements in this way.  Instead, we
+must expose most of the underlying logic of the `TransitionExecutor`, and task
+the users with building their own exploration strategies with these primitives.
+
+It is likely that we will be able to provide some higher-level functionality to users
+by way of wrapper libraries we implement on top of the proposed API, but that
+work should be left to a subsequent iteration.
+
+**NOTE:** This interface is intended as an abstract API, to communicate the
+mappings from request to reply. See the [gRPC Example](#grpc-example) for a
+sketch of what the actual implementation may look like.
+
+```scala
+/** A State is a map from variable names to values, and represents a concrete state.  */
+type State = Map[String, TlaEx]
+
+/** An abstract representation of a transition.
+ *
+ * These correspond to the numbered transitions that the `TransitionExectur` uses
+ * to advance frames. But we likely want to present some more illuminating metadata
+ * associated with the transitions to the users. E.g., ability to view a
+ * representation as a `TlaEx` or see an associated operator name (if any)?  */
+type Transition
+
+/** An execution is an alternating sequence of states and transitions,
+ *  terminating in a state with no transition.
+ *
+ * It is a high-level representation of the `EncodedExecution` maintained by the
+ * transition executor).
+ *
+ * E.g., an execution from `s1` to `sn` with transitions `ti` for each `i` in `1..n-1`:
+ * 
+ *     List((s1, Some(t1)), (s2, Some(t2)), ..., (sn, None))
+ */
+type Execution = List[(State, Option[Transition])]
+
+trait UninitializedModel = {
+  val spec: TlaModule
+  val transitions: Set(Transition)
```

Comment:
Shall we have a sequence of transitions instead of a set? It's often important to know the order of the transitions.